src/repl.ml

```diff
@@ -0,0 +1,152 @@
+open Import
+
+let name = "REPL"
+
+let has_utop sandbox =
+  let open Promise.Syntax in
+  let cmd =
+    let bin =
+      match Platform.t with
+      | Win32 -> "where"
+      | Darwin
+      | Linux
+      | Other ->
+        "which"
+    in
+    Sandbox.get_command sandbox bin [ "utop" ]
+  in
+  let+ result = Cmd.output cmd in
+  match result with
+  | Error _ -> false
+  | Ok _ -> true
+
+let can_build sandbox =
+  let open Promise.Syntax in
+  let includes = `String "**/{dune-project}" in
+  let* dunes = Workspace.findFiles ~includes () in
+  match dunes with
+  | [] -> Promise.return false
+  | _ :: _ :: _ ->
+    (* If the workspace contains several dune-project, we don't know how to
+       build it. *)
+    Promise.return false
+  | el :: _ -> (
+    let cwd = Filename.dirname (Uri.fsPath el) |> Path.of_string in
+    let cmd = Sandbox.get_command sandbox "dune" [ "build" ] in
+    let+ result = Cmd.output ~cwd cmd in
+    match result with
+    | Error _ -> false
+    | Ok _ -> true )
+
+let create_terminal instance sandbox =
+  match Extension_instance.repl instance with
+  | Some term ->
+    Terminal_sandbox.show term;
+    Promise.return (Ok term)
+  | None -> (
+    let open Promise.Syntax in
+    let* has_utop = has_utop sandbox in
+    let* can_build = can_build sandbox in
+    let cmd =
+      match (has_utop, can_build) with
+      | true, true -> Sandbox.get_command sandbox "dune" [ "utop" ]
```

Comment:
What would be simpler than all of this is to just ask the user what command to use to launch the repl.

Reply:
We would still have to detect if utop is available and handle the potential errors. I'm not sure asking the user would simplify the code that much

Reply:
Sorry, I meant simplify from the point of the user. Currently, this functionality doesn't really work if the user wants to use `down`, `rtop`, or just pass some flags to `dune utop`. Also, it requires remembering a few implementation details about the project that aren't too intuitive.  If we let the user pick how they launch the toplevel we can sidestep all of this.

Reply:
I see, maybe we could have a setting for the command used to run the repl? This way users would have a good default, and they could change their repl if needed.

I think most users would be happy with utop, as it's the recommended repl for OCaml, so I wouldn't want to interrupt the workflow with a popup for them.

Reply:
I added settings to set the REPL path and arguments. Here are the new settings:

| Name                               | Description                                                                                             | Default |
| ---------------------------------- | ------------------------------------------------------------------------------------------------------- | ------- |
| `ocaml.repl.path.linux`            | The path of the REPL that the extension uses on Linux                                                   | `null`  |
| `ocaml.repl.path.osx`              | The path of the REPL that the extension uses on macOS                                                   | `null`  |
| `ocaml.repl.path.windows`          | The path of the REPL that the extension uses on Windows                                                 | `null`  |
| `ocaml.repl.args.linux`            | The REPL arguments that the extension uses on Linux                                                     | `null`  |
| `ocaml.repl.args.osx`              | The REPL arguments that the extension uses on macOS                                                     | `null`  |
| `ocaml.repl.args.windows`          | The REPL arguments that the extension uses on Window                                                    | `null`  |

Reply:
Why is it per platform? I get that it might be useful in some cases, but most users only use a single platform.

Reply:
Agree it's not very useful, I removed them in favor of `ocaml.repl.path` and `ocaml.repl.args`